source/common/filesystem/kqueue/watcher_impl.cc

```diff
@@ -0,0 +1,170 @@
+  }
+
+  FileWatchPtr watch(new FileWatch());
+  watch->fd_ = watch_fd;
+  watch->file_ = path;
+  watch->events_ = events;
+  watch->callback_ = cb;
+  watch->watching_dir_ = watching_dir;
+
+  int flags = NOTE_DELETE | NOTE_RENAME;
+  if (watching_dir) {
+    flags = NOTE_DELETE | NOTE_WRITE;
+  }
+
+  struct kevent event;
+  EV_SET(&event, watch_fd, EVFILT_VNODE, EV_ADD | EV_CLEAR, flags, 0,
+         reinterpret_cast<void*>(watch_fd));
+
+  if (kevent(queue_, &event, 1, NULL, 0, NULL) == -1) {
+    throw EnvoyException(
+        fmt::format("unable to add filesystem watch for file {}: {}", path, strerror(errno)));
+  }
+
+  if (event.flags & EV_ERROR) {
+    throw EnvoyException(
+        fmt::format("unable to add filesystem watch for file {}: {}", path, strerror(event.data)));
+  }
+
+  ENVOY_LOG(debug, "added watch for file: '{}' fd: {}", path, watch_fd);
+
+  watches_[watch_fd] = watch;
+
+  return watch;
+}
+
+void WatcherImpl::removeWatch(FileWatchPtr& watch) {
+  int fd = watch->fd_;
+  watches_.erase(fd);
+}
+
+void WatcherImpl::onKqueueEvent() {
+  struct kevent event = {};
+  timespec nullts = {0, 0};
+
+  while (true) {
+    uint32_t events = 0;
+    int nevents = kevent(queue_, NULL, 0, &event, 1, &nullts);
+    if (nevents < 1 || event.udata == nullptr) {
+      return;
+    }
+
+    int watch_fd = reinterpret_cast<std::intptr_t>(event.udata);
+
+    FileWatchPtr file = watches_[watch_fd];
+    if (file == nullptr) {
+      continue;
+    }
+    ASSERT(watch_fd == file->fd_);
+
+    if (file->watching_dir_) {
+      if (event.fflags & NOTE_DELETE) {
+        // directory was deleted
+        removeWatch(file);
+        return;
+      }
+
+      if (event.fflags & NOTE_WRITE) {
+        // directory was written -- check if the file we're actually watching appeared
+        FileWatchPtr new_file = addWatch_(file->file_, file->events_, file->callback_, true);
+        if (new_file != nullptr) {
+          removeWatch(file);
+          file = new_file;
+
+          events |= Events::MovedTo;
+        }
+      }
+    } else {
+      // kqueue doesn't seem to work well with NOTE_RENAME and O_SYMLINK, so instead if we
+      // get a NOTE_DELETE on the symlink we check if there is another file with the same
+      // name we assume a NOTE_RENAME and re-attach another event to the new file.
+      if (event.fflags & NOTE_DELETE) {
+        removeWatch(file);
+
+        FileWatchPtr new_file = addWatch_(file->file_, file->events_, file->callback_, true);
+        if (new_file == nullptr) {
+          return;
+        }
+
+        event.fflags |= NOTE_RENAME;
+        file = new_file;
+      }
+
+      if (event.fflags & NOTE_RENAME) {
+        events |= Events::MovedTo;
+      }
+    }
+
+    if (events & file->events_) {
+      ENVOY_LOG(debug, "matched callback: file: {}", file->file_);
+      file->callback_(events);
+    }
+
+    ENVOY_LOG(debug, "notification: fd: {} flags: {:x} file: {}", file->fd_, event.fflags,
```

Comment:
Given the callback may trigger a bunch of logs on its own, you may want to bump this log line up before so you get the log about the event on the file before all the things triggered by that event.